2025-10-21-feature-flags-recurring-outages.md

```diff
@@ -0,0 +1,269 @@
+- **Critical monitoring gap**: CPU alerting was completely missing throughout these incidents, preventing early detection of CPU pressure that was the root cause of three outages. Added comprehensive CPU alerting to catch CPU pressure before it causes service degradation.
+- **Observability improvements**: Added monitoring for previously invisible failure modes
+
+### Short-term Improvements ([Tracked in GitHub Issue #40885](https://github.com/PostHog/posthog/issues/40885))
+
+**In progress (next 2 weeks):**
+
+- **Strike team formation**: Engineers from flags, ingestion, and infrastructure teams conducting comprehensive review of application and infrastructure to identify remaining bottlenecks
+- **Redis isolation**: Investigating decoupling flags Redis instance from application Redis instance to prevent cross-service impact
+
+**To complete before re-enabling ArgoCD sync:**
+
+- Evaluate current state of synced flags deployment and ensure durability against future outages
+- Update flags service charts config to match values currently in ArgoCD
+- Define deployment strategy for short-term (considering deployment vs rollout to avoid 503s)
+- Define and implement Redis strategy
+- Establish feature flags team as hard code-owners for flags-related code
+
+### Medium-term Improvements
+
+**Incident response and monitoring:**
+
+- Build high-level dashboard of important flag metrics with runbook links
+- Implement rollout/annotation controls to disable staged rollouts and enable "force-merge" for rolling changes
+- Update feature flag runbooks with dashboard links and deeper investigation paths
+- Add missing alerts against existing service/infrastructure level metrics
+- Update readiness checks to validate dependencies that degraded under load (e.g., ping database instead of mirroring liveness checks)
+
+**Architectural improvements:**
+
+- **Rate limiting for cache operations**: Prevent Redis overwhelm from cache population
+- **Connection pool monitoring**: Automatic throttling when pools approach exhaustion
+- **Connection limiting**: Prevent unbounded concurrent connections
+
+### Long-term Improvements
+
+- **Load testing framework**: Production-scale testing to catch load-dependent issues before deployment
+- **Progressive rollout infrastructure**: Gradual deployments to limit blast radius
+- **Deployment strategy evolution**: Re-evaluate rollout vs deployment approaches with programmatic controls
+- **Comprehensive monitoring**: Document Postgres-to-Redis data flow patterns and create runbooks for data transfer storm scenarios
+
+## Lessons Learned
+
+### What Went Well
+
+✅ **Rapid detection** – Monitoring caught issues within 2 minutes in most cases  
+✅ **Quick initial response** – Rollbacks executed immediately when possible  
+✅ **Systematic investigation** – Teams methodically identified overload patterns  
+✅ **Clear communication** – Customer updates throughout incidents  
```

Comment:
Should the timelines include the Customer updates? Like when we notified customers via status page etc.?